ext/ptrace/ptrace.c

```diff
@@ -21,7 +23,18 @@ ptrace_peekdata(VALUE _mod, VALUE pid, VALUE addr, VALUE _data) {
 
 static VALUE
 ptrace_peekuser(VALUE _mod, VALUE pid, VALUE off, VALUE _data) {
-  long val = ptrace(PTRACE_PEEKUSER, NUM2LONG(pid), NUM2LONG(off), NULL);
+  // aarch64 doesn't support PTRACE_PEEKUSER, but it does support
+  // PTRACE_GETREGSET which is fairly similar. Since i386 and x86_64 support it
+  // too, just use that unconditionally to present the same API for the
+  // registers we care about.
+  size_t offset = NUM2LONG(off);
+  long val;
+  size_t size = offset + sizeof(val);
+  char *regs = malloc(size);
```

Comment:
check the return value is not NULL

Reply:
Or maybe directly declare `struct user_pt_regs regs`

Reply:
Yeah so the code here is kind of strange, perhaps I should have added a comment explaining why I'm not using `struct user_pt_regs`. The TL;DR is that I couldn't find a header that declared it consistently: my x86_64 Ubuntu machine has `struct user_regs_struct` declared in /usr/include/x86_64-linux-gnu/sys/user.h, but the aarch64 Alpine (musl-based) rootfs I was using has `struct user_pt_regs` in /usr/include/asm/ptrace.h. The untyped, dynamically-sized buffer is my way of getting around this by not using that struct at all, although it has the safety issues you've outlined. I'm open to suggestions as to how you think I should do this :)

Reply:
Oh I didn't notice this. Maybe we can check the source of GDB to see how it handles this issue, though I guess it simply uses if-def guards to perform arch-dependent operations.

Anyways, in this case it's fine to use malloc/free here, but please add a comment on why we can't use `struct user_regs_struct` or similar structures here.

And for the validation of `offset`, set a reasonable bound (4096) is enough (but also add a comment why we can't compare it with `sizeof(...)`). I just don't want an invalid input from Ruby could lead the program crash.